tiledb/sm/filesystem/vfs.cc

```diff
@@ -362,18 +363,27 @@ Status VFS::filelock_lock(const URI& uri, filelock_t* fd, bool shared) const {
   // Hold the lock while updating counts and performing the lock.
   std::unique_lock<std::mutex> lck(filelock_mtx_);
 
-  // Only need to actually lock the file if this is the first one on the URI.
-  if (incr_lock_count(uri)) {
+  auto it = process_filelocks_.find(uri.to_string());
+  if (it != process_filelocks_.end()) {
+    it->second.first++;
+    // we need to return the lock for the xlock semantics
+    *lock = it->second.second;
     return Status::Ok();
   }
 
-  if (uri.is_file())
+  // We must hold the fd in the global map in order to free from any context
+  if (uri.is_file()) {
 #ifdef _WIN32
-    return win_.filelock_lock(uri.to_path(), fd, shared);
+    auto st = win_.filelock_lock(uri.to_path(), lock, shared);
 #else
-    return posix_.filelock_lock(uri.to_path(), fd, shared);
+    auto st = posix_.filelock_lock(uri.to_path(), lock, shared);
 #endif
 
+    if (st.ok())
+      process_filelocks_[uri.to_string()] = {1, *lock};
```

Comment:
@stavrospapadopoulos @tdenniston I wanted to highlight this to double-check, because it is a change from the previous sequencing. I believe that previously a URI could be written to the global counter but would not be decremented if the locking failed; with this change to check the status, we should only write to the global counter if the lock call succeeds.

Reply:
I noticed the change - it is the right thing to do, I had omitted that.